Add TypeScript support

sypl opened the pull request:
First attempt at adding TypeScript support. Based of propType comments and own experiences using lib.

Notes:
- The description for onSuccess and onExit and their actual callback argument(s) don't match, as per this discussion: https://github.com/plaid/react-native-plaid-link-sdk/issues/18. Callback args reflect what's actually coming through.
- I'm not sure which fields are nullable.
- ~~Exporting these enums to be used in .ts files doesn't seem to work, they don't get compiled to objects (at least for me)~~ Fixed by using literal type over enum

Kyonru commented on package.json:
```diff
@@ -28,7 +28,8 @@
     "react": "^16.8.1",
-    "react-native": "^0.61.3"
+    "react-native": "^0.61.3",
+    "react-native-gesture-handler": "^1.6.0"
```
👀

Kyonru replied:
Is this really needed?

sypl replied:
Oh you know what, no. Sorry, somehow thought that TouchableOpacity was only from that package.

sypl replied:
Okay, fixed.

sypl commented:
@Kyonru Are the types for fields all ok? I've marked some as nullable, which I've done because they actually came back as null for me (admittedly only in dev mode), but I'm not sure if other ones are. The published schema (https://plaid.com/docs/#onsuccess-callback) don't seem to denote any fields as nullable, even though some have come back null. I'm thinking now perhaps I should do away with nulls altogether.

Also the callback comments do not match up with the signatures in terms of number of arguments. I've taken the comments directly from the PlaidLink.js file, but I'm thinking that it's better to amend them so that they say callback takes a single metadata argument.

sypl commented:
I made a few changes, need a re-review I'm afraid @Kyonru:
- In absence of certainty about nullable fields, I've removed them entirely. I know some of mine were wrong anyway. For example,`request_id` on onExit is always included (https://plaid.com/docs/#request_id)
- I changed some names to be more in line with API docs, so EventName over EventType and Product over ProductType

sypl commented:
@Kyonru Hey sorry, could I get a rereview of this? Thanks.

alexmbp commented on react-native-plaid-link-sdk.d.ts:
```diff
@@ -0,0 +1,203 @@
+    name: string;
+    subtype: string;
+    type: string;
+  };
```
I think it might be a good idea to export it as a separate interface. In that case we will be able to use it in `accounts` array also (extending with verification_status).

sypl replied:
Thanks for the tip, I've incorporated the changes.

Kyonru commented:
@sypl Sorry, I'm being sick last couple of days. I agree with @alexmbp.
Also, we are not contributors, maybe @wkiefer or @jwoogerd can help us with getting this approved.

sypl commented:
> @sypl Sorry, I'm being sick last couple of days. I agree with @alexmbp.
> Also, we are not contributors, maybe @wkiefer or @jwoogerd can help us with getting this approved.

That's fair, @Kyonru . @wkiefer @jwoogerd, does this addition seem acceptable?

Diego-Franco commented:
Hi @sypl do you know if plaid has ETA about this PR?

sypl commented:
@Diego-Franco No, afraid not. I have no affiliation with Plaid - I just added this as a suggestion. Waiting now on the contributors @wkiefer @jwoogerd

safaiyeh commented:
Would love this as well, could someone have a look @wkiefer

kinigitbyday commented on react-native-plaid-link-sdk.d.ts:
```diff
@@ -0,0 +1,199 @@
+  export interface PlaidLinkProps<
+    C = TouchableOpacity,
+    P = TouchableOpacityProps
+  > {
```
I've typed pieces of this library in the past for work, and you should be able to simplify this a bit so it doesn't need two generic arguments, and also allows enforcing that the type of the component props actually matches for the component. I ended up with something like:

```
export interface PlaidLinkProps<T> {
  ...otherProps,
  component?: (props: T) => React.ReactElement;
  componentProps?: T;
}

class PlaidLink<T> extends Component<PlaidLinkProps<T>> {}
export default PlaidLink;
```

Might even be able to push it a bit further and require that `T` extends something like `{ onPress: () => void }`, because otherwise it'll be unclickable (which is somewhat common sense, but not clear from the documentation here).

Maftalion commented on react-native-plaid-link-sdk.d.ts:
```diff
@@ -0,0 +1,199 @@
+  timestamp: string;
+}
+
+export interface OnEventArgs {
+  eventName: EventName;
+  metadata: EventMetadata;
+}
```
are you getting these? I just get an object with `event` and `metadata` as properties.

StoyanD commented:
this looks really good, can you guys review and merge it?

zsweigart commented:
Hi @sypl Thanks for the contribution, we have just released version 7.0.0-rc1 which now supports typescript.  We welcome any feedback you have on the new version. thanks!

sypl commented:
@zsweigart Good news! 2020 isn't a complete loss! 😄
